### PR TITLE
Eas test

### PR DESCRIPTION
**Pull request by tynes**
For code review

### CODE REVIEW

File: packages/contracts-bedrock/test/vendor/eas/SchemaRegistry.t.sol [hunk 1]
```diff
@@ -0,0 +1,290 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.19;
+
+import { Test } from "forge-std/Test.sol";
+import { console } from "forge-std/console.sol";
+import { SchemaRegistry, SchemaRecord } from "src/vendor/eas/SchemaRegistry.sol";
+import { ISchemaResolver } from "src/vendor/eas/resolver/ISchemaResolver.sol";
+
+contract SchemaRegistryTest is Test {
```
Comment (tynes) [inline packages/contracts-bedrock/test/vendor/eas/SchemaRegistry.t.sol]: You should do `contract SchemaRegistryTest is CommonTest` and then update the following code: https://github.com/ethereum-optimism/optimism/blob/a47441c8995e1b02fcddbbe558c65be3763a8a7f/packages/contracts-bedrock/test/setup/Setup.sol#L112

so that `schemaRegistry` and `eas` are defined and set in the `Setup` file. Then they will be available globally here, so you don't need to deploy the `registry`. This tests the L2 genesis script, so the same setup that creates an L2 genesis for production is tested in the unit test. The other unit tests do this

File: packages/contracts-bedrock/test/vendor/eas/SchemaRegistry.t.sol [hunk 2]
```diff
@@ -0,0 +1,290 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.19;
+
+import { Test } from "forge-std/Test.sol";
+import { console } from "forge-std/console.sol";
```
Comment (tynes) [inline packages/contracts-bedrock/test/vendor/eas/SchemaRegistry.t.sol:5]: `console` should be deleted before this PR is finalized

File: packages/contracts-bedrock/test/vendor/eas/SchemaRegistry.t.sol [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+        return keccak256(abi.encodePacked(schema, resolver, revocable));
+    }
+
+    // Should be equal to the current version of SchemaRegistry.sol in this repository
+    function testVersion() public view {
+        assertEq(registry.version(), "1.3.1-beta.1");
+    }
+
+    // Basic functionality tests
+    function testRegisterSchema() public {
```
Comment (tynes) [inline packages/contracts-bedrock/test/vendor/eas/SchemaRegistry.t.sol]: You can turn this into a fuzz test rather than hardcoding values, ie:

```solidity
    function testRegisterSchema(string memory _schema, address _resolver, bool _revocable) public {
```

then you can delete the hardcoded values. This improves the test coverage, foundry will run the test a ton of times (based on fuzz runs config in `foundry.toml`) so now it covers many values rather than just some values. If there is a value that doesn't work, you use `vm.assume(bool)` to filter out values, see the foundry docs and other uses of `vm.assume` in the codebase

File: packages/contracts-bedrock/test/vendor/eas/EAS.t.sol [hunk 4]
```diff
@@ -0,0 +1,2758 @@
...
+    bytes32 constant ZERO_BYTES32 = bytes32(0);
+    bytes32 private constant ATTEST_TYPEHASH =
+        0xfeb2925a02bae3dae48d424a0437a2b6ac939aa9230ddc55a1a76f065d988076;
+    bytes32 private constant REVOKE_TYPEHASH =
+        0x4e1c85c87bc4c1867b4225cc3fb634a4e0fd8a91feb1ebca195aeaf6611a773b;
+
+    // =============================================================
+    //                        ERROR SELECTORS
+    // =============================================================
+    bytes4 constant InvalidRegistrySelector =
```
Comment (tynes) [inline packages/contracts-bedrock/test/vendor/eas/EAS.t.sol]: This is the same as `error InvalidRegistry()` - its better to use the higher level syntax rather than the hash + `bytes4`, the compiler turns it into the `bytes4` under the hood

File: packages/contracts-bedrock/test/vendor/eas/EAS.t.sol [hunk 5]
```diff
@@ -0,0 +1,2758 @@
...
+        senderKey = 0x12345;
+        sender = vm.addr(senderKey);
+
+        // Setup accounts
+        sender2 = makeAddr("sender2");
+        recipient = makeAddr("recipient");
+        recipient2 = makeAddr("recipient2");
+
+        // Create registry at a temporary address
+        registry = new SchemaRegistry();
```
Comment (tynes) [inline packages/contracts-bedrock/test/vendor/eas/EAS.t.sol]: If we use `CommonTest` like noted in the other comment, we don't need to deploy the `registry` here

File: packages/contracts-bedrock/test/setup/Setup.sol [hunk 6]
```diff
@@ -111,7 +113,9 @@ contract Setup {
     ISuperchainTokenBridge superchainTokenBridge = ISuperchainTokenBridge(Predeploys.SUPERCHAIN_TOKEN_BRIDGE);
     IOptimismSuperchainERC20Factory l2OptimismSuperchainERC20Factory =
         IOptimismSuperchainERC20Factory(Predeploys.OPTIMISM_SUPERCHAIN_ERC20_FACTORY);
-
+    IEAS eas = IEAS(Predeploys.EAS);
```
Comment (tynes) [inline packages/contracts-bedrock/test/setup/Setup.sol:117]: This is exactly correct

File: packages/contracts-bedrock/test/vendor/eas/SchemaRegistry.t.sol [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+    ISchemaRegistry registry;
+
+    // Events
+    event Registered(
+        bytes32 indexed uid,
+        address indexed registerer,
+        SchemaRecord schema
+    );
+
+    function setUp() public override {
```
Comment (tynes) [inline packages/contracts-bedrock/test/vendor/eas/SchemaRegistry.t.sol:27]: You don't need to define `registry` or call `setUp`, the `setUp` function in `CommonTest` will be called and `schemaRegistry` will exist as a global variable from your change in `Setup.sol`

File: packages/contracts-bedrock/test/vendor/eas/SchemaRegistry.t.sol [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+        bytes32 uid = registry.register(_schema, ISchemaResolver(_resolver), _revocable);
+        SchemaRecord memory record = registry.getSchema(uid);
+        
+        assertEq(record.uid, uid);
+        assertEq(record.schema, _schema);
+        assertEq(address(record.resolver), _resolver);
+        assertEq(record.revocable, _revocable);
+    }
+
+    function testGetSchema() public {
```
Comment (tynes) [inline packages/contracts-bedrock/test/vendor/eas/SchemaRegistry.t.sol]: This should be a fuzz test as well

File: packages/contracts-bedrock/test/vendor/eas/SchemaRegistry.t.sol [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+        assertEq(record.uid, uid);
+        assertEq(record.schema, schema);
+        assertEq(address(record.resolver), resolver);
+        assertEq(record.revocable, revocable);
+    }
+
+    // Edge cases
+    function testRegisterSchemaWithoutSchema() public {
+        string memory schema = "";
+        address resolver = address(0x123);
```
Comment (tynes) [inline packages/contracts-bedrock/test/vendor/eas/SchemaRegistry.t.sol]: This should be a fuzz test for resolver and revocable but hardcode `""` for schema

File: packages/contracts-bedrock/test/vendor/eas/SchemaRegistry.t.sol [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+
+    // Error cases
+    function testCannotRegisterSameSchemaTwice() public {
+        string memory schema = "bool isFriend";
+        address resolver = address(0);
+        bool revocable = true;
+
+        registry.register(schema, ISchemaResolver(resolver), revocable);
+
+        vm.expectRevert(abi.encodeWithSignature("AlreadyExists()"));
```
Comment (tynes) [inline packages/contracts-bedrock/test/vendor/eas/SchemaRegistry.t.sol]: The cleaner way to do this:

```solidity
// define this under `SchemaRegistryTest`
error AlreadyExists();
```

`vm.expectRevert(AlreadyExists.selector)`

File: packages/contracts-bedrock/test/vendor/eas/SchemaRegistry.t.sol [hunk 11]
```diff
@@ -0,0 +1,512 @@
...
+    ///         - Verifies Registered event emission
+    ///         - Validates all schema parameters:
+    ///           * UID matches expected
+    ///           * Schema string correct
+    ///           * Resolver address matches
+    ///           * Revocable flag set properly
+    ///      Ensures proper event emission and data accuracy
+    ///      during schema registration
+    function testRegisterSchemaEvent() public {
+        string memory schema = "bool like";
```
Comment (tynes) [inline packages/contracts-bedrock/test/vendor/eas/SchemaRegistry.t.sol]: Anywhere you hardcode inputs like this you should turn into a fuzz test

File: packages/contracts-bedrock/test/vendor/eas/eip1271/EIP1271Verifier.t.sol [hunk 12]
```diff
@@ -0,0 +1,411 @@
...
+    // =============================================================
+    //                      BASIC STATE TESTS
+    // =============================================================
+    /// @dev Tests initial verifier configuration
+    function testInitialState() public view {
+        assertEq(verifier.getName(), "EAS");
+        assertEq(verifier.getNonce(signer), 0);
+    }
+
+    // =============================================================
```
Comment (tynes) [inline packages/contracts-bedrock/test/vendor/eas/eip1271/EIP1271Verifier.t.sol]: These comments are nice, its generally better to match the style of the codebase than diverge even if the divergence is better

File: packages/contracts-bedrock/test/vendor/eas/resolver/SchemaResolver.t.sol [hunk 13]
```diff
@@ -0,0 +1,414 @@
...
+    }
+
+    /// @dev Returns fixed version string
+    function version() external pure returns (string memory) {
+        return "1.3.0";
+    }
+}
+
+/// @dev Minimal mock EAS implementation for testing
+contract MockEAS {
```
Comment (tynes) [inline packages/contracts-bedrock/test/vendor/eas/resolver/SchemaResolver.t.sol]: I recommend replacing this `MockEAS` contract with `vm.mockCall`

See https://book.getfoundry.sh/cheatcodes/mock-call?highlight=mockCall#mockcall

You can basically do:

```solidity
vm.mockCall(
  address(eas),
  abi.encodeWithSelector(IEAS.attest.selector),
  abi.encode(true)
)
```

Then when you call `eas.attest`, it will return `true`

This will remove the need to have mock contracts and you can do this `vm.mockCall` inside of `setUp` which will replicate you overwriting `eas` with `MockEAS`

File: packages/contracts-bedrock/test/vendor/eas/SchemaRegistry.t.sol [hunk 14]
```diff
@@ -0,0 +1,512 @@
...
+
+        bytes32 expectedUID = _getUID(schema, resolver, revocable);
+        SchemaRecord memory expectedSchema = SchemaRecord({
+            uid: expectedUID,
+            schema: schema,
+            resolver: ISchemaResolver(resolver),
+            revocable: revocable
+        });
+
+        vm.expectEmit(true, true, true, true, address(registry));
```
Comment (tynes) [inline packages/contracts-bedrock/test/vendor/eas/SchemaRegistry.t.sol]: smol nit: you can do `vm.expectEmit(address(registry))` and its the same thing as using all of the true values, less complicated that way

File: packages/contracts-bedrock/test/vendor/eas/EAS.t.sol [hunk 15]
```diff
@@ -0,0 +1,3493 @@
...
+        
+        // Initialize test variables
+        senderKey = 0x12345;
+        sender = vm.addr(senderKey);
+        sender2 = makeAddr("sender2");
+        recipient = makeAddr("recipient");
+        recipient2 = makeAddr("recipient2");
+
+        // Get contracts from predeploys
+        registry = ISchemaRegistry(Predeploys.SCHEMA_REGISTRY);
```
Comment (tynes) [inline packages/contracts-bedrock/test/vendor/eas/EAS.t.sol]: This should be defined already in `Setup.sol`

File: packages/contracts-bedrock/test/vendor/eas/EAS.t.sol [hunk 16]
```diff
@@ -0,0 +1,3476 @@
...
+
+    /// @dev Forwards attestation request to the verifying contract
+    function attestByDelegation(
+        DelegatedAttestationRequest calldata request
+    ) external returns (bytes32) {
+        return IEAS(verifyingContract).attestByDelegation(request);
+    }
+}
+
+// Mainnet Contract
```
Comment (tynes) [inline packages/contracts-bedrock/test/vendor/eas/EAS.t.sol:68]: nit: I recommend using natspec comments: https://docs.soliditylang.org/en/latest/natspec-format.html

File: packages/contracts-bedrock/test/vendor/eas/EAS.t.sol [hunk 17]
```diff
@@ -0,0 +1,3476 @@
...
+contract EASTest is CommonTest {
+    // =============================================================
+    //                           CONSTANTS
```
Comment (tynes) [inline packages/contracts-bedrock/test/vendor/eas/EAS.t.sol]: You should delete these big spacers, will be more likely to be merged, none of the upstream code has these

**Comment by tynes** [comment]
Looks like this needs to be rebased on `develop` because of a merge conflict. 

`git pull --rebase <remote> <branch>`

Depending on your remote names, it could be:

`git pull --rebase origin develop`

After that, you can force push to this branch. Then it looks good to be opened against the main optimism repo. I do think its a bit large tho, if you copy/pasted tests only for a subset to a new branch, that may work best, like only tests for EAS rather than all of them. Then after that is merged we can work on getting the next contracts tests merged
